cluster/v1alpha1/types.go

```diff
@@ -157,3 +160,230 @@ var ReservedClusterClaimNames = [...]string{
+// placement namespace.
+type PlacementSpec struct {
+	// Predicates represent a slice of predicates to select ManagedClusters. The predicates are ORed.
+	// +optional
+	Predicates []ClusterPredicate `json:"predicates,omitempty"`
+
+	// Affinity represents the scheduling constraints. It will be applied to all predicates.
+	// +optional
+	Affinity *Affinity `json:"affinity,omitempty"`
+}
+
+// ClusterPredicate represents a predicate to select ManagedClusters.
+type ClusterPredicate struct {
+	// ClusterSets represent the ManagedClusterSets from which the ManagedClusters are selected.
+	// If the slice is empty, ManagedClusters will be selected from the ManagedClusterSets bound to the placement
+	// namespace, otherwise ManagedClusters will be selected from the intersection of this slice and the
+	// ManagedClusterSets bound to the placement namespace.
+	// +optional
+	ClusterSets []string `json:"clusterSets,omitempty"`
+
+	// NumberOfClusters represents the number of ManagedClusters to be selected which meet the
+	// requirements in this predicate.
+	// 1) If not specified, it indicates all matched ManagedClusters will be selected;
+	// 2) Otherwise if the nubmer of matched ManagedClusters is larger than NumberOfClusters, a random subset
+	//    with expected number of ManagedClusters will be selected;
+	// 3) If the nubmer of matched ManagedClusters is equal to NumberOfClusters, all matched ManagedClusters
+	//    will be selected;
+	// 4) If the nubmer of matched ManagedClusters is less than NumberOfClusters, all matched ManagedClusters
+	//    will be selected, and the status of condition `PlacementConditionSatisfied` will be set to false;
+	// +optional
+	NumberOfClusters *int32 `json:"numberOfClusters,omitempty"`
+
+	// RequiredDuringSchedulingRequiredDuringExecution represents a scheduling strategy for a
+	// cluster selector. If specified, the cluster selector should be matched at both scheduling
+	// time (when the placement decisions are made for the first time) and execution time (After
+	// one or multiple placement decisions are made). It means,
+	// 1) Any ManagedCluster, which does not match the selector, should not be selected;
+	// 2) If a selected ManagedCluster ceases to match the selector (e.g. due to an update) after the
+	//    placement decisions are made, it will be eventually removed from the placement decisions;
+	// 3) If a ManagedCluster starts to match the selector after the placement decisions are made, it
+	//    will either be selected or at least has chance to be selected (when NumberOfClusters is
+	//    specified);
+	//
+	// Other scheduling strategies not yet implemented includes
+	// RequiredDuringSchedulingIgnoredDuringExecution and PreferredDuringSchedulingIgnoredDuringExecution
+	// +optional
+	RequiredDuringSchedulingRequiredDuringExecution *ClusterSelector `json:"requiredDuringSchedulingRequiredDuringExecution,omitempty"`
```

Comment:
These are logically ANDed with the `clusterSets` if specified, right?

Reply:
Yes, it is.